lib/src/relations/to_many.dart

```diff
@@ -0,0 +1,271 @@
+import 'dart:collection';
+import 'dart:ffi';
+
+import 'package:meta/meta.dart';
+import 'package:objectbox/src/bindings/bindings.dart';
+import 'package:objectbox/src/bindings/helpers.dart';
+
+import 'info.dart';
+import '../box.dart';
+import '../modelinfo/entity_definition.dart';
+import '../store.dart';
+
+/// Manages a to-many relation, an unidirectional link from a "source" entity to
+/// multiple objects of a "target" entity.
+///
+/// You can:
+///   - [add()] new objects to the relation.
+///   - [removeAt()] target objects from the relation at a specific list index.
+///   - [remove(id:)] target objects from the relation by an ID.
+///
+/// ```
+/// class Student {
+///   final teachers = ToMany<Teacher>();
+///   ...
+/// }
+///
+/// // Example 1: create a relation
+/// final teacher1 = Teacher();
+/// final teacher2 = Teacher();
+///
+/// final student1 = Student();
+/// student1.teachers.add(teacher1);
+/// student1.teachers.add(teacher2);
+///
+/// final student2 = Customer();
+/// student2.teachers.add(teacher2);
+///
+/// // saves students as well as teachers in the database
+/// store.box<Student>().putMany([student1, student2]);
+///
+///
+/// // Example 2: remove a relation
+///
+/// student.teachers.removeAt(index)
+/// student.teachers.applyToDb(); // or store.box<Student>().put(student);
+/// ```
+class ToMany<EntityT> extends Object with ListMixin<EntityT> {
+  /*late final*/ Store _store;
+
+  /*late final*/
+  Box<EntityT> _box;
+
+  /*late final*/
+  Box _srcBox;
+
+  /*late final*/
+  EntityDefinition<EntityT> _entity;
+
+  RelInfo _rel;
+
+  List<EntityT> /*?*/ __items;
+  final _counts = <EntityT, int>{};
+
+  @override
+  int get length => _items.length;
+
+  @override
+  set length(int newLength) {
+    if (newLength < _items.length) {
+      _items
+          .getRange(newLength, _items.length)
+          .forEach((EntityT element) => _track(element, -1));
+    }
+    _items.length = newLength;
+  }
+
+  @override
+  EntityT operator [](int index) => _items[index];
+
+  @override
+  void operator []=(int index, EntityT element) {
+    final items = _items;
+    if (0 > index || index >= items.length) {
+      throw RangeError.index(index, this);
+    }
+    if (index < items.length) {
+      if (items[index] == element) return;
+      _track(items[index], -1);
+    }
+    items[index] = element;
+    _track(element, 1);
+  }
+
+  @override
+  void add(EntityT element) {
+    if (element == null) ArgumentError.notNull('element');
+    // TODO add and addAll() don't need to load from DB, unless already loaded
```

Comment:
To do? I guess `List<EntityT> get _items => __items ??= _loadItems();` takes care of this?

Reply:
no, that actually loads items from DB. The TODO is about not loading from DB at all if you just add to the list. There's no need because we don't match against existing items anyway and let core take care of duplicate additions.

Consider this code:

```dart
final src = box.get(1);
src.relMany.add()/addAll()
box.put(src);
```

In such cases, we don't have to load the original items in the list, we can just try to add new ones. This can be a significant improvement for relations with many items.

Reply:
I've implemented it after all.

Reply:
LGTM!